### PR TITLE
Add background modifier

### PR DESCRIPTION
**Pull request by supernintendo**
Adds support for the [background](https://developer.apple.com/documentation/swiftui/view/background(alignment:content:)) modifier and adds the ability to use namespaced child elements when decoding modifiers. This is based on @carson-katri's idea for handling nested views within modifiers via Content Entities as described [here](https://github.com/liveview-native/liveview-client-swiftui/issues/170).

```heex
<VStack id="background">
  <Spacer />
  <HStack id="example_heart">
    <Image system-name="heart.fill" modifiers={@native |> background(alignment: :center, content: :heart_bg) |> foreground_style(primary: {:color, :white})}>
      <background:heart_bg>
        <Circle modifiers={@native |> frame(width: 32, height: 32) |> foreground_style(primary: {:color, :red})} />
      </background:heart_bg>
    </Image>
  </HStack>
  <Spacer />
  <HStack id="example_liveviewnative">
    <Text modifiers={@native |> background(alignment: :center, content: :text_bg) |> foreground_style(primary: {:linear_gradient, %{gradient: {:colors, [:orange, :indigo, :purple]}, start_point: {0, 0}, end_point: {1, 1}}})}>
      liveviewnative
      <background:text_bg>
        <RoundedRectangle corner-radius="8" modifiers={@native |> frame(width: 192, height: 64)} />
      </background:text_bg>
    </Text>
  </HStack>
  <Spacer />
</VStack>
```

Screenshot 2023-03-23 at 5 58 14 PM

There are a few outstanding issues with this PR:

1. The `foreground_style` modifier influences the background content instead of the `background_style` modifier which doesn't appear to do anything. I'm not sure why this is but from what I can tell the native [backgroundStyle modifier](https://developer.apple.com/documentation/swiftui/grid/backgroundstyle(_:)) only affects nested content rendered using the `in:` constructor and not content passed to `content`.

2. There seem to be performance limitations with nested modifier views in my implementation. By rendering too many elements with background modifiers within the same view I'm able to produce a crash, although I haven't investigated this enough to know what might be causing it.

3. ~The string format for `content` when referencing namespaced child views could be a bit confusing. I'd like to tweak the Elixir modifier schema so that it accepts an atom instead, which should ameliorate this.~ Fixed in [2ccb51e](https://github.com/liveview-native/liveview-client-swiftui/pull/734/commits/2ccb51e8dde56862f6fbdee23503e2d304b682e1)

Closes https://github.com/liveview-native/liveview-client-swiftui/issues/183.

### CODE REVIEW

File: Sources/LiveViewNative/ViewTree.swift [hunk 1]
```diff
@@ -93,14 +93,17 @@ struct ViewTreeBuilder<R: RootRegistry> {
...
     }
     
-    private func applyModifiers(encoded: String?, to view: some View, context: LiveContextStorage<R>) -> some View {
+    private func applyModifiers(encoded: String?, to view: some View, element: ElementNode, context: LiveContextStorage<R>) -> some View {
         let modifiers: [ModifierContainer<R>]
         if let encoded {
             let decoder = JSONDecoder()
+            decoder.userInfo[.liveContext!] = context
+            decoder.userInfo[.elementNode!] = element
+
```
Comment (shadowfacts) [inline Sources/LiveViewNative/ViewTree.swift:106]: The context and element shouldn't be stored and passed through the decoder, they can be accessed from the modifier using their respective property wrappers:
```swift
struct MyModifier<R: RootRegistry>: ViewModifier, Decodable {
    @ObservedElement private var element
    @LiveContext<R> private var context
    // etc.
}
```

File: Sources/LiveViewNative/Registries/BuiltinRegistry.swift [hunk 2]
```diff
@@ -16,10 +16,7 @@ protocol BuiltinRegistryProtocol {
     static func decodeModifier(_ type: ModifierType, from decoder: Decoder) throws -> BuiltinModifier
 }
 
-struct BuiltinRegistry: BuiltinRegistryProtocol {
-    
-    static let attributeDecoder = JSONDecoder()
-    
+struct BuiltinRegistry<R: RootRegistry>: BuiltinRegistryProtocol {
```
Comment (shadowfacts) [inline Sources/LiveViewNative/Registries/BuiltinRegistry.swift:19]: This makes sense, but can you also remove the `R` generic type and the `context` parameter from the `lookup` method, for consistency?

**Review by shadowfacts — CHANGES_REQUESTED** [review]
Something is definitely weird with how the background/foreground modifiers are interacting, though it's not obvious to me what the issue is.

In plain old SwiftUI, the following (translated directly from your example) results in just a red circle appearing, since the foreground style is applying to both the Circle and the Image.
```swift
Image(systemName: "heart.fill")
	.background(alignment: .center) {
		Circle()
			.frame(width: 32, height: 32)
	}
	.foregroundStyle(.red)
```

To get the result that you actually show, you'd need to move the foreground style _inside_ the background modifier so that it only applies to the Circle:
```swift
Image(systemName: "heart.fill")
	.background(alignment: .center) {
		Circle()
			.frame(width: 32, height: 32)
			.foregroundStyle(.red)
	}
```

**Comment by carson-katri** [comment]
`Image` does not respect the `foreground_style` modifier because it uses a `foregroundColor` modifier in its body, which uses the system foreground color when set to `nil`. The example below produces the same result as the provided screenshot:

```swift
Image(systemName: "heart.fill")
            .foregroundColor(nil)
            .background(alignment: .center) {
                Circle()
                    .frame(width: 32, height: 32)
            }
            .foregroundStyle(.red)
```

https://github.com/liveview-native/liveview-client-swiftui/blob/931a824c85f17fc2e85924fde82b1d3d23cee6ad/Sources/LiveViewNative/Views/Images/Image.swift#L45

The attribute should probably be removed from `Image`.

Text has [the same problem](https://github.com/liveview-native/liveview-client-swiftui/blob/931a824c85f17fc2e85924fde82b1d3d23cee6ad/Sources/LiveViewNative/Views/Text%20Input%20and%20Output/Text.swift#L78). The `textColor`, `font`, and `fontWeight` attributes should likely be replaced with modifiers as well.

**Comment by carson-katri** [comment]
RE the `background_style` modifier, this would require using the [`BackgroundStyle`](https://developer.apple.com/documentation/swiftui/shapestyle/background) shape style. Background content created in this way uses the foreground style by default.

**Comment by shadowfacts** [comment]
Ah, that is indeed the case. We do probably want general modifiers for them, but, I think the `foregroundColor`, `font`, and `fontWeight` modifiers need to remain attributes on the Text view as well, since doing them through our usual modifiers architecture would break composing Texts with different styles.

I think a workaround would be having a helpers like `foregroundColorIfPresent` for Text that only applies the underlying modifier if the value is non-nil, rather than applying `.foregroundColor(nil)` as currently happens.

File: Sources/LiveViewNative/Modifiers/BackgroundModifier.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
+//
+//  BackgroundModifier.swift
+// LiveViewNative
+//
+//  Created by May Matyi on 3/17/23.
+//
+
+import SwiftUI
+
+struct BackgroundModifier<R: RootRegistry>: ViewModifier, Decodable {
```
Comment (carson-katri) [inline Sources/LiveViewNative/Modifiers/BackgroundModifier.swift:10]: Could you document the modifier?

Comment (supernintendo) [reply Sources/LiveViewNative/Modifiers/BackgroundModifier.swift]: @carson-katri Done!

**Comment by supernintendo** [comment]
> 

I added `foregroundColorIfPresent` to fix that in [f7e7172](https://github.com/liveview-native/liveview-client-swiftui/pull/734/commits/f7e71724f2ecf6fa8f02e7610e241cf6e5449173). Let me know if there are other examples of this I should address, although maybe that's out of scope for this pull request?

**Review by carson-katri — APPROVED** [review]

**Review by shadowfacts — CHANGES_REQUESTED** [review]
This looks good, just needs a slight tweak to make the docs visible

**Review by shadowfacts — APPROVED** [review]

**Pull request merged**